airbyte-integrations/connectors/source-shopify/source_shopify/shopify_graphql/bulk/job.py

```diff
@@ -227,15 +266,50 @@ def job_retry_on_concurrency(self, request: requests.PreparedRequest) -> Union[A
             return self.job_retry_concurrent(request)
 
     def job_healthcheck(self, response: requests.Response) -> Optional[requests.Response]:
-        # errors check
-        errors = self.job_check_for_errors(response)
-        # when the concurrent job takes place, we typically need to wait and retry, but no longer than 10 min.
-        if not self.has_running_concurrent_job(errors):
-            return response if not errors else None
-        else:
-            # get the latest request to retry
-            request: requests.PreparedRequest = response.request
-            return self.job_retry_on_concurrency(request)
+        # get the latest request to retry
+        request: requests.PreparedRequest = response.request
+        try:
+            errors = self.job_check_for_errors(response)
+            # when the concurrent job takes place, we typically need to wait and retry, but no longer than 10 min.
+            if not self.has_running_concurrent_job(errors):
+                return response if not errors else None
+            else:
+                return self.job_retry_on_concurrency(request)
+        except (ShopifyBulkExceptions.BulkJobBadResponse, ShopifyBulkExceptions.BulkJobUnknownError) as err:
+            # sometimes we face with `HTTP-500 Internal Server Error`
+            # we should retry such at least once
+            return self.job_retry_request(request)
+
+    def should_expand_slice_size(self) -> bool:
+        return self.job_should_expand_slice_size and not self.job_should_reduce_slice_size
+
+    def should_reduce_slice_size(self) -> bool:
+        return not self.job_should_expand_slice_size and self.job_should_reduce_slice_size
+
+    def _expand_slice_size(self) -> None:
+        self.job_should_expand_slice_size = True
+        self.job_should_reduce_slice_size = False
+        # adjust the expand factor
+        # `0.5` is the coef. to guarantee the positive increse, which should be noticible
+        self.slice_size_expand_factor = (
+            0.5 + (self.job_elapsed_time_threshold_sec - self.current_job_elapsed_time) / self.job_elapsed_time_threshold_sec
+        )
+
+    def _reduce_slice_size(self) -> None:
+        self.job_should_expand_slice_size = False
+        self.job_should_reduce_slice_size = True
+
+    def adjust_slice_size(self) -> None:
+        if self.current_job_elapsed_time > self.job_elapsed_time_threshold_sec:
+            self._reduce_slice_size()
+        elif self.current_job_elapsed_time < 1 or self.current_job_elapsed_time < self.job_last_elapsed_time_sec:
```

Comment:
Once the time needed for the job to COMPLERE is less than 1, less than the previous job finished - we expand.